### DISCUSSION TITLE
On the computation of the sensitivities

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#18784] [Q&A General] [answered]
Hi 
This topic is a follow-up to the topic in #18464 where we seek a way to compute the gradient of a cost functional w.r.t material parameter. 
Let's consider the diffusion equation with spatial heterogenous diffusivity, the final discretized form would be of the form `A(m) u=b` where m is the material parameters (the diffusivity). 
To invert for the diffusivities `m`, given measured data `d` , the reconstruction process seeks minimizing a cost functional, which can be written as `J(m)=0.5 (Pu-d)^T (Pu-d)`, where `P` is a projection operator to project `u` to the location of the measurements `d`.

The `ith` element in the gradient vector `g_i = ∂J(m)/∂m_i = (∂u/∂m_i)^T (Pu-d) ` . This means, for `M` number of parameters, computing `g` requires `M` evaluation of `(∂u/∂m_i )` which means `M` forward solves. This is very expensive especially when the problem is parameterized such that each finite element has a different material parameter (i.e. the diffusivity in this example).

One way to circumvent this is to expand `(∂u/∂m_i)`, this yields `g_i=u^T  (∂A/∂m_i)^T v `with `v=(A^T)^(-1) P^T (d-Pu(m))`. This means, we only need one forward solve to compute `v`, then we need to evaluate the sparse matrix-vector multiplication`(∂A/∂m_i)^T v` which will result in a sparse vector to be inner-product with `u` in  `g_i=u^T  (∂A/∂m_i)^T v `. 

So, we no longer need `M` forward solves and just need one. Now the whole effort is to get `u^T (∂A/∂m_i)^T v ` within moose. 
I have some plan and I need evaluation, and/or suggestions on how to improve it.
 1-Solve for u, then use the results to solve for v. 
 2- Have a map of diffusivity_vs._element, where each diffusivity parameter id has a pointer to the corresponding finite element.
 3- In a third input file, have auxkernel to postprocess`u^T  (∂A/∂m_i)^T v ` where the material is defined as constant function = 1 **because we only have the diffusivity and we took the derivative with respect to it**.
 4- repeat 3 for all M paramters!
Computing `(∂A/∂m_i)^T v ` is not clear for me since we need to compute it for a given diffusivity_id and repeat that for all diffusivities `M` . 

Any hint specifically on the mapping part (parameter id vs. elements) and on how to post-process in point 3?

**Comment by lynnmunday** [comment]
@aaelmeli 
In (1), is v the solution to the adjoint equation?  I would use something like a the dirac kernel reporterPointSource to apply the projection in your adjoint solve.  You can read your misfit into this using a csvReader vector postprocessor.  

Can you do material inversion for the much simpler case where the entire block has a single unknown diffusivity?  I would start with this since it will require a new auxkernel to compute the gradient term mentioned in #18464    Once that works, you could make a more complicated problem where the diffusivity is a parameterized function of some spatial dimension?  

I think an optimization problem where every element has a different diffusivity will be hard to solve and will require Tikhonov regularization.  I'm not sure how you can create this map of element to diffusivity.  I think @Traiwit has asked about a similar capability, maybe they can give you some advice on having a material property that is different in each element in this disucssion #18727

If you were going to take approach 4, we have the simplex algorithm in Tao working.  We are in the process of getting our optimization code approved for release.  I think this will take another month but at that point you could use and develop that code.  

Are you going to use something like python to do your optimization and do the data transfers using files?

**Comment by aaelmeli** [comment]
> In (1), is v the solution to the adjoint equation? 

yes, it is.

> Can you do material inversion for the much simpler case where the entire block has a single unknown diffusivity? I would start with this since it will require a new auxkernel to compute the gradient term mentioned in #18464 Once that works, you could make a more complicated problem where the diffusivity is a parameterized function of some spatial dimension?

Yes, that is probably better and what I should start with. In this regard, I had some questions, 

a-Regarding the input file/s, based on what you suggested in your reply on (1), we can compute `u` and `v` using _two_ different input files, Am I correct? 
then to compute  `u^T (∂A/∂m_i)^T v` , we need to have another **additional input file** (so we would have 3 input files to compute g_i), Am I correct?

b- Do you mean by new auxkernel that I need to implement a new auxkernel that takes u and v from previous run as input, and compute `u^T (∂A/∂m_i)^T v`? or do we have something to reuse in moose?

c- In the input file where we want to compute `u^T (∂A/∂m_i)^T v`, since this term is just matrix_vector multiplication and inner product of the result with `u`, so there is no solve is required, my question now, is it possible to have dummy kernel and dummy variable and explicitly tell moose not to solve the linear system of equations?

> I think an optimization problem where every element has a different diffusivity will be hard to solve and will require Tikhonov regularization. I'm not sure how you can create this map of element to diffusivity. I think @Traiwit has asked about a similar capability, maybe they can give you some advice on having a material property that is different in each element in this disucssion #18727 

I have implemented this approach for 2D wave-equation inversion in frequency domain and the results were quite well. Now, I am using moose in order to do a similar thing but in 3D efficiently. 
Regarding parameter_vs_elements map, I have design my in-house code to have a map of `parameterId_vs_elements`, for example, given the parameter_id, it returns a vector of pointers to the finite elements associated with that parameter. However, because of my little experience with moose, I was not sure if this was possible or not, maybe they have something more elegant and more efficient.

> If you were going to take approach 4, we have the simplex algorithm in Tao working. We are in the process of getting our optimization code approved for release. I think this will take another month but at that point, you could use and develop that code.

That will be very helpful especially when I do wave equation inversion in 3D.

> Are you going to use something like python to do your optimization and do the data transfers using files?

Yes, my initial thought is to have something working to test a research idea and not worry much about the efficiency. So, I am going to link Matlab with moose. I will use Matlab's optimization routines for the optimization and communicate with moose through files. these files are: one contains the parameters at the `kth` iteration, the other one contains the value of the cost function and the gradient vector (computed in moose).

**Comment by lynnmunday** [comment, marked as answer]
I've done something similar to this where I was using matlab to do the optimization. I've attached two input files to this that I used for the forward and adjoint input files.  These input fils were used for for load inversion for a point source so it is was much simpler than the material inversion problem because the adjoint variable was my gradient, I didn't need to compute a gradient term.  

To solve u^T (∂A/∂m_i)^T v    I think you will run the forward problem, save the primary variable (u) to an exodus file and then read that exodus file into the adjoint problem, read u into a auxvariable using a userObject.  Then in the adjoint problem, use a parsedAuxKernel to multiply the adjoint variable, v, by u.  Then use a postprocessor to sum all of this up across all of the elements.

I used something like this for a convective boundary condition where temperature_forward is the solution field from teh forward problem:

```
[AuxVariables]
  [temperature_forward]
  []
  [T2]
  []
[]
[AuxKernels]
  [TT]
    type = ParsedAux
    args = 'temperature temperature_forward'
    variable = T2
    function = 'temperature*(100-temperature_forward)'
  []
[]

[Variables]
  [temperature]
  []
[]

[Postprocessors]
  [adjoint_pt_0]
    type = SideIntegralVariablePostprocessor
    variable = T2
    boundary = left
  []
[]
```

I saw this discussion https://github.com/idaholab/moose/discussions/18790
and they are computing material properties on one app and passing them to the subapp.  Maybe if you grep for ADMaterialRealAux you might find a place where it is being used in a material.  something like that might work in place of the material to element map you need.

**Comment by lynnmunday** [comment]
This is the forward input file:
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 1.0
    ymin = 0
    ymax = 2.0
    elem_type = QUAD4
    nx = 10
    ny = 20
  []
[]

[Variables]
  [temperature]
  []
[]

[Kernels]
  [heat_conduction]
    type = ADHeatConduction
    variable = temperature
  []
[]

[DiracKernels]
  [force_0]
    type = ReporterPointSource
    variable = 'temperature'
    value_name = 'csv_reader/value'
    x_coord_name = csv_reader/x
    y_coord_name = csv_reader/y
    z_coord_name = csv_reader/z
  []
[]

[BCs]
  [top]
    type = DirichletBC
    variable = temperature
    boundary = top
    value = 0
  []
  [bottom]
    type = DirichletBC
    variable = temperature
    boundary = bottom
    value = 0
  []
  [left]
    type = DirichletBC
    variable = temperature
    boundary = left
    value = 0
  []
  [right]
    type = DirichletBC
    variable = temperature
    boundary = right
    value = 0
  []
[]

[Materials]
  [steel]
    type = ADGenericConstantMaterial
    prop_names = thermal_conductivity
    prop_values = .5
  []
[]

[Problem]
  type = FEProblem
[]

[Executioner]
  type = Steady
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[VectorPostprocessors]
  [temperatures]
    type = NodalValueSampler
    variable = temperature
    block = '0'
    sort_by = id
    outputs = fullResponseVector
  []
  [csv_reader]
    type = CSVReader
    csv_file = 'zForwardInput/inputForces.csv'
  []
[]

[Outputs]
  [exodus]
    file_base = 'zForwardOutput/out'
    type = Exodus
    execute_on = final
  []
  [fullResponseVector]
    file_base = 'zForwardOutput/all'
    type = CSV
    execute_vector_postprocessors_on = final
  []
[]

```

**Reply by lynnmunday** [reply]
I had to paste them in, I couldn't figure out how to attach them.

**Comment by lynnmunday** [comment]
this is the adjoint input file:
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 1.0
    ymin = 0
    ymax = 2.0
    elem_type = QUAD4
    nx = 10
    ny = 20
  []
[]

[Variables]
  [temperature]
  []
[]

[Kernels]
  [heat_conduction]
    type = ADHeatConduction
    variable = temperature
  []
[]

[DiracKernels]
  [force_0]
    type = ReporterPointSource
    variable = temperature
    value_name = csv_reader/value 
    x_coord_name = csv_reader/x
    y_coord_name = csv_reader/y
    z_coord_name = csv_reader/z
  []
[]

[BCs]
  [top]
    type = DirichletBC
    variable = temperature
    boundary = top
    value = 0
  []
  [bottom]
    type = DirichletBC
    variable = temperature
    boundary = bottom
    value = 0
  []
  [left]
    type = DirichletBC
    variable = temperature
    boundary = left
    value = 0
  []
  [right]
    type = DirichletBC
    variable = temperature
    boundary = right
    value = 0
  []
[]

[Materials]
  [steel]
    type = ADGenericConstantMaterial
    prop_names = thermal_conductivity
    prop_values = .5
  []
[]

[Problem]
  type = FEProblem
[]

[Executioner]
  type = Steady
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[VectorPostprocessors]
  [temperatures]
    type = NodalValueSampler
    variable = temperature
    block = '0'
    sort_by = id
    outputs = fullResponseVector
  []
  [csv_reader]
    type = CSVReader
    csv_file = 'zAdjointInput/inputForces.csv'
    # execute_on = initial
  []
[]

[Outputs]
  [exodus]
    file_base = 'zAdjointOutput/out'
    type = Exodus
    execute_on = final
  []
  [fullResponseVector]
    file_base = 'zAdjointOutput/all'
    type = CSV
    execute_vector_postprocessors_on = final
  []
[]

```
